src/utils.hpp

```diff
@@ -0,0 +1,132 @@
+#ifndef UTILS_H
+#define UTILS_H
+#include <deal.II/base/conditional_ostream.h>
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/base/tensor_function.h>
+
+#include <deal.II/distributed/fully_distributed_tria.h>
+
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/dofs/dof_tools.h>
+#include <deal.II/grid/grid_tools.h>
+
+#include <deal.II/fe/fe_simplex_p.h>
+#include <deal.II/fe/fe_system.h>
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/fe/fe_values_extractors.h>
+
+#include <deal.II/grid/grid_in.h>
+
+#include <deal.II/lac/solver_cg.h>
+#include <deal.II/lac/trilinos_precondition.h>
+#include <deal.II/lac/trilinos_sparse_matrix.h>
+
+#include <deal.II/numerics/data_out.h>
+#include <deal.II/numerics/matrix_tools.h>
+#include <deal.II/numerics/vector_tools.h>
+
+#include <deal.II/base/work_stream.h>
+#include <deal.II/base/multithread_info.h>
+
+
+using namespace dealii;
+
+// N is the number of gating variables, 3 in our case
+template<int N>
+struct GatingVariables {
+    double& get(int k) {
+        //assert(k < N);
+        return var[k];
+    }
+private:
+    double var[N];
+};
+
+// tissue conductivity tensor
+template <int dim>
+class D : public TensorFunction<2, dim>
+{
+public:
+    D() : TensorFunction<2, dim>() {}
+    void value_list(const std::vector<Point<dim>> &/*points*/, std::vector<Tensor<2, dim>> & /*values*/) const override {} // do not use it
+
+    typename TensorFunction<2, dim>::value_type value(const Point<dim> & /*p*/) const override {
+        Tensor<2, dim> sigma = unit_symmetric_tensor<dim>();
+        //return sigma * 1.171e2;
+        sigma[0][0] = sigma_il * sigma_el / (sigma_il + sigma_el);
+        sigma[1][1] = sigma_it * sigma_et / (sigma_it + sigma_et);
+        sigma[2][2] = sigma_it * sigma_et / (sigma_it + sigma_et);
+        return sigma;
+    }
+
+private:
+    double sigma_il = 0.17; /*S/m*/
```

Comment:
```suggestion
    const double sigma_il = 0.17; /*S/m*/
```
or even better
```suggestion
    static constexpr double sigma_il = 0.17; /*S/m*/
```
(and same for the others).